### DISCUSSION TITLE
Precedence between two "text" constructs

### DISCUSSION THREAD

**Discussion by StaloneLab** [micromark/micromark#80] [Q&A] [answered]
Hello :wave: 

I maintain the ZMarkdown repository, which has a lot of extensions to remark. I have so far almost managed to get `remark-kbd` switched to micromark, but I struggle to find how to give precedence to my custom `kbd` element over `emphasis`. The problem specified here only concern micromark by itself, remark is not involved.

The code for the extension can be found here: https://github.com/zestedesavoir/zmarkdown/tree/next/packages/micromark-extension-kbd , and here is a test, that will not give what I expect it to:

```markdown
*foo||*||
```

gives

```html
<p><em>foo||</em>||</p>
```

but I would want it to give:

```html
<p>*foo<kbd>*</kbd></p>
```

As you can see, currently, `attention` has a priority over `kbd`. I have not found how it would be possible to invert this behavior, excepted by switching `kbd` from *text* to *flow*, which doesn't seem like a good solution. Could you indicate me how I could manage to get it working as expected?

Thanks a lot :smiley:

**Comment by wooorm** [comment, marked as answer]
hey hey!

That conflicts a bit with markdown:

```markdown
*a~b*c~

~a*b~c*
```

Yields:

*a~b*c~

~a*b~c*

So the earlier one “wins”.
That’s because emphasis/strong/delete/link/image are parsed separately: as potential starts/ends, and finally when done resolved based on whether they match and which comes first, and the starts/ends that don’t match reverted back to text.

But, `kbd` could perhaps be parsed similarly to `code` / [`math`](https://github.com/micromark/micromark-extension-math)? So that it parses from start to end in one go.

And then its contents chunks could be linked to form more text? That would be similar to [titles](https://github.com/micromark/micromark/blob/main/packages/micromark-factory-title/dev/index.js) (but then using text instead of string as the content type)

**Reply by StaloneLab** [reply]
Thanks for your reply!

I had indeed looked at `code`, which worked as expected, but hadn't noticed why.
What you suggest works very well.

I'm not sure however of what you mean by the last sentence: "its contents chunks could be linked to form more text".
Could you explain a bit more?

**Reply by wooorm** [reply]
depends on what you want:
- Do character references and character entities work in kbd?
- Do emphasis and links and such work in kbd?

**Reply by StaloneLab** [reply]
Inline constructs do not work inside `kbd`, so no emphasis, links, or anything. I'm not sure about character references, but it seems to me like we do not want them to work in `kbd` (as in `code`? I'm unsure about it).

I'm curious though, how would you make another construct work inside `kbd`? Should I call the tokenizer from inside?

**Reply by wooorm** [reply]
I was assuming you did want such things, hence these questions and the “And then its contents chunks could be linked to form more text? That would be similar to titles (but then using text instead of string as the content type)”

- If you don‘t want anything inside, take code as inspiration
- If you want character references, take `title`
- If you want all phrasing constructs, take `title` but use `text` instead of `string` as a content type.

Adding a `contentType` on a token, and linking them together, will be seen by micromark as something to parse in the future. When you’re done with what you parsed, micromark will recursively find that embedded content, and parse it

**Reply by wooorm** [reply]
did you see https://github.com/micromark/micromark#creating-a-micromark-extension, that might help clarify such things.
And, you can also look at the types in the code. It’s quite extensive: https://github.com/micromark/micromark/blob/main/packages/micromark-util-types/index.js

**Reply by StaloneLab** [reply]
I knew the document existed, but hadn't taken time to read it until now; it answers my questions, thanks.

Also, hats off for the work on `micromark`, seems very powerful :+1:
